compactor/src/compact.rs

```diff
@@ -167,47 +240,399 @@ impl Compactor {
+        // Nothing to compact
+        if overlapped_files.is_empty() {
+            return Ok(None);
+        }
 
-impl TablePartition {
-    fn new(sequencer_id: SequencerId, table_id: TableId, partition_id: PartitionId) -> Self {
-        Self {
-            sequencer_id,
-            table_id,
-            partition_id,
+        // One file without tombstone, no need to compact
+        if overlapped_files.len() == 1 && overlapped_files[0].tombstones.is_empty() {
+            return Ok(None);
         }
-    }
-}
 
-struct ParquetFileWithTombstone {
-    data: Arc<ParquetFile>,
-    tombstones: Vec<Tombstone>,
-}
+        // Keep the fist IoxMetadata to reuse same IDs and names
+        let iox_metadata = overlapped_files[0].iox_metadata();
+
+        // Verify if the given files belong to the same partition
+        // Note: we can ignore this verification if we assume this is a must-have condition
+        if overlapped_files.len() > 1 {
+            for file in &overlapped_files {
+                if file.data.sequencer_id != overlapped_files[0].data.sequencer_id
+                    || file.data.table_id != overlapped_files[0].data.table_id
+                    || file.data.partition_id != overlapped_files[0].data.partition_id
+                {
+                    return Err(Error::NotSamePartition {
+                        sequencer_id_1: overlapped_files[0].data.sequencer_id,
+                        table_id_1: overlapped_files[0].data.table_id,
+                        partition_id_1: overlapped_files[0].data.partition_id,
+                        sequencer_id_2: file.data.sequencer_id,
+                        table_id_2: file.data.table_id,
+                        partition_id_2: file.data.partition_id,
+                    });
+                }
+            }
+        }
 
-impl ParquetFileWithTombstone {
-    fn tombstone_ids(&self) -> HashSet<TombstoneId> {
-        self.tombstones.iter().map(|t| t.id).collect::<HashSet<_>>()
+        // Convert the input files into QueryableBatches for making query plan
+        let query_chunks: Vec<_> = overlapped_files
+            .iter()
+            .map(|f| {
+                f.to_queryable_parquet_chunk(
+                    Arc::clone(&self.object_store),
+                    iox_metadata.table_name.to_string(),
+                    iox_metadata.partition_key.to_string(),
+                )
+            })
+            .collect();
+
+        // Compute min & max sequence numbers
+        let mut min_sequence_number = query_chunks[0].min_sequence_number();
+        let mut max_sequence_number = query_chunks[0].max_sequence_number();
+        for c in &query_chunks {
+            if min_sequence_number > c.min_sequence_number() {
+                min_sequence_number = c.min_sequence_number();
+            }
+            if max_sequence_number < c.max_sequence_number() {
+                max_sequence_number = c.max_sequence_number();
+            }
+        }
```

Comment:
Same idea as above, plus using the existing `min` and `max` functions

```suggestion
        let (head, tail) = query_chunks.split_first().unwrap();
        let mut min_sequence_number = head.min_sequence_number();
        let mut max_sequence_number = head.max_sequence_number();
        for c in tail {
            // use std::cmp::{min, max};
            min_sequence_number = min(min_sequence_number, c.min_sequence_number());
            max_sequence_number = max(max_sequence_number, c.max_sequence_number());
        }
```